### PR TITLE
ginko custom reporter to azure app insights

### PR DESCRIPTION
**Pull request by mjudeikis**
```release-note
NONE
```

builds on top of https://github.com/openshift/openshift-azure/pull/1365

Enables us to view test result in summaries view in App insights

### CODE REVIEW

**Comment by mjudeikis** [comment]
Go to: 
`Azure Portal -> ResourceGroups -> fakerp-upgrades-insights -> osa-fakeRP-upgrades -> Analytics`
and execute:
```
customEvents
| extend componentText = extractjson('$.ComponentTexts', name)
| extend failed = extractjson('$.Failed', name)
| extend failureMessage = extractjson('$.FailureMessage', name)
| where timestamp > ago(30m) 
| summarize count(componentText) by failed, componentText
| render barchart
```

File: docs/e2e/appinsights.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+		"Passed":         specSummary.Passed(),
+		"Skipped":        specSummary.Skipped(),
+	}
+```
+
+## Example queries
+
+Results can be queries using Kusto queries below:
+
+Get all Passed & failed test count in last 30 minutes
```
Comment (thekad) [inline docs/e2e/appinsights.md]: Passed should be lowercase

File: env.example [hunk 2]
```diff
@@ -36,6 +36,9 @@ export DEPLOY_VERSION=v3.11
 # into OpenShift
 #export AZURE_AAD_GROUP_ADMINS_ID=9e4f44bd-90a1-44b2-b89a-ed9884662958
 
+# Azure App insights key for monitoring shim to send data
+# export AZURE_APP_INSIGHTS_KEY=
```
Comment (thekad) [inline env.example]: this insights key... should we add it to secrets?

File: pkg/util/blackbox/types.go [hunk 3]
```diff
@@ -22,6 +25,7 @@ type Blackbox interface {
 // Config defines blackbox instance configuration
 type Config struct {
 	Cli              *http.Client
+	Icli             insights.TelemetryClient
```
Comment (thekad) [inline pkg/util/blackbox/types.go]: nit: ICli

File: env.example (hunk 2 above)
Comment (mjudeikis) [reply env.example]: Already added

**Comment by mjudeikis** [comment]
https://github.com/openshift/openshift-azure/issues/1362
/retest

File: test/reporters/appinsights.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/onsi/ginkgo/types"
+
+	"github.com/openshift/openshift-azure/pkg/util/azureclient/insights"
+)
+
+type AzureAppInsightsReporter struct {
+	icli insights.TelemetryClient
+}
+
+func NewAzureAppInsightsReporter() *AzureAppInsightsReporter {
```
Comment (jim-minter) [inline test/reporters/appinsights.go]: unexport AzureAppInsightsReporter, return interface type gingko.Reporter

File: test/reporters/appinsights.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	icli := insights.NewTelemetryClient(os.Getenv("AZURE_APP_INSIGHTS_KEY"))
+	icli.Context().CommonProperties["type"] = "gotest"
```
Comment (jim-minter) [inline test/reporters/appinsights.go]: suggest s/gotest/ginkgo/ - we will probably want another reporter like this for `go test` in the future

File: pkg/util/azureclient/insights/insights.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package insights
```
Comment (jim-minter) [inline pkg/util/azureclient/insights/insights.go]: this is all confused - I suggest not bothering creating pkg/util/azureclient/insights at all

File: test/clients/azure/client.go [hunk 7]
```diff
@@ -47,6 +48,7 @@ type Client struct {
 	VirtualNetworks                  azureclient.VirtualNetworksClient
 	VirtualNetworksPeerings          azureclient.VirtualNetworksPeeringsClient
 	Groups                           azureclient.GroupsClient
+	Insights                         insights.TelemetryClient
```
Comment (jim-minter) [inline test/clients/azure/client.go]: not used - don't add it

File: cmd/monitoring/monitoring.go [hunk 8]
```diff
@@ -36,12 +37,17 @@ var (
...
 
 	resourceGroup  string
 	subscriptionID string
 
 	instances []instance
+
+	// cluster creation e2e time
+	startTime time.Time
+	stopTime  time.Time
+	duration  time.Time
```
Comment (jim-minter) [inline cmd/monitoring/monitoring.go]: duration is not used

File: cmd/monitoring/monitoring.go [hunk 9]
```diff
@@ -238,6 +254,22 @@ func loadOCConfig() (*api.OpenShiftManagedCluster, error) {
 	return cs, nil
 }
 
+func (m *monitor) trackCreationTimestamp() {
```
Comment (jim-minter) [inline cmd/monitoring/monitoring.go]: I don't think this stuff is right - running monitoring != creating a cluster
we can add this kind of instrumentation in the fakerp separately

File: pkg/util/azureclient/insights/insights.go (hunk 6 above)
Comment (mjudeikis) [reply pkg/util/azureclient/insights/insights.go]: Its not part of main azure sdk. And still in development. This way I split it from production code (?)

Comment (jim-minter) [reply pkg/util/azureclient/insights/insights.go]: why not import github.com/Microsoft/ApplicationInsights-Go/appinsights direct when you need it? what's the purpose of this package?

**Comment by thekad** [comment]
bad rebase?

**Comment by jim-minter** [comment]
/retest

@thekad upstream go-bindata broke, we had to pin it in our build image

File: pkg/util/azureclient/insights/insights.go (hunk 6 above)
Comment (mjudeikis) [reply pkg/util/azureclient/insights/insights.go]: trying to be smart here :D not working very well :D

**Pull request merged**